src/fighter_state.rs

```diff
@@ -0,0 +1,747 @@
+
+                // If it was not an additive transition
+                if !was_additive {
+                    // Skip processing other transitions because our current state was removed, and
+                    // return true to indicate that a non-additive transition has been performed.
+                    return true;
+                }
+            }
+        }
+
+        // I we got here we are still in the same state so return false to indicate no non-additive
+        // transitions have been performed.
+        false
+    }
+}
+
+/// Component on fighters containing the action intent queue.
+///
+/// Actions are things that the fighter should try to do that doesn't represent a state change. The
+/// current fighter state must choose to act upon an action intent for it to be effect, and it will
+/// otherwise be ignored.q
+#[derive(Component, Default, Deref, DerefMut)]
+pub struct ActionIntents(VecDeque<Box<dyn Reflect>>);
+
+//
+// Fighter state components
+//
+
+/// Component indicating the player is idling
+#[derive(Component, Reflect, Default, Debug)]
+#[component(storage = "SparseSet")]
+pub struct Idling;
+impl Idling {
+    pub const PRIORITY: i32 = 0;
+    pub const ANIMATION: &'static str = "idle";
+}
+
+/// Component indicating the player is moving
+#[derive(Component, Reflect, Default, Debug)]
+#[component(storage = "SparseSet")]
+pub struct Moving {
+    pub velocity: Vec2,
+}
+impl Moving {
+    pub const PRIORITY: i32 = 10;
+    pub const ANIMATION: &'static str = "running";
+}
+
+/// The player is throwing an item
+#[derive(Component, Reflect, Default, Debug)]
+pub struct Throwing;
+impl Throwing {
+    pub const PRIORITY: i32 = 15;
+}
+
+/// The player is grabbing an item ( or trying to)
+#[derive(Component, Reflect, Default, Debug)]
+pub struct Grabbing;
+impl Grabbing {
+    pub const PRIORITY: i32 = Throwing::PRIORITY;
+}
+
+/// Component indicating the player is flopping
+#[derive(Component, Reflect, Default, Debug)]
+#[component(storage = "SparseSet")]
+pub struct Attacking {
+    /// The initial y-height of the figther when starting the attack
+    pub start_y: f32,
+    pub has_started: bool,
+    pub is_finished: bool,
+}
+impl Attacking {
+    pub const PRIORITY: i32 = 30;
+    pub const ANIMATION: &'static str = "attacking";
+}
+
+/// Component indicating the player is getting knocked back
+#[derive(Component, Reflect, Default, Debug)]
+#[component(storage = "SparseSet")]
+pub struct KnockedBack {
+    pub velocity: Vec2,
+    pub timer: Timer,
+}
+impl KnockedBack {
+    pub const PRIORITY: i32 = 20;
+    pub const ANIMATION_LEFT: &'static str = "knocked_left";
+    pub const ANIMATION_RIGHT: &'static str = "knocked_right";
+}
+
+/// Component indicating the player is dying
+#[derive(Component, Reflect, Default, Debug)]
+#[component(storage = "SparseSet")]
+pub struct Dying;
+impl Dying {
+    pub const PRIORITY: i32 = 1000;
+    pub const ANIMATION: &'static str = "dying";
+}
```

Comment:
Since this moves towards a more dynamic fighter state list than before, the idea is that we don't necessarily know all of the possible animations that might exist at compile time, so each state is responsible for asking the fighter to play a named animation, instead of a fixed enum variant.

We'll probably want to be somewhat standardized on which animations should exist as a baseline for fighters, but this will allow mods to create new fighters with completely new states, attacks, and animations.

Reply:
I think I'm sort pointing toward my perception that the dynamic states stuff could have been separated from the rest of the state refactor since enabling that was "part of scripting improvements" in my mind, but I think ultimately I agree with the design goal and it implicates the state refactor just as much as it does scripting so I think I dont really have a reason to object.

Reply:
Yeah, in this case I think what was essential for scripting, still turns out useful to us for easily extending our own design as we move on, as long as our end design goal isn't to only ever have a fixed set of animations.

Here is one middle-ground alternative, worth considering, though:

We have a core set of animations we expect to be implemented for every fighter, we could enumerate those, but we allow custom, arbitrarily named animations also, that could be triggered by custom attacks. That way you could try to make sure every fighter implements the "Core" animations, but it wouldn't stop mods from adding new animations.

That could be good to help allow flexibility, while still enforcing some sort of standard. And having a standard is good for other things, too, for instance, if we have special effects later. If we required that all fighters were required to have the "stunned" core animation, then any time somebody invents a new mod with a new status effect like "stuck in a bubble", they would know that it would be safe to play the "stunned" animation, because all fighters have the "stunned" animation.

But if somebody made a new fighter, it would still be possible for them to have the "ultimate_smash" animation, which is non-standard and only available to that fighter.

Reply:
Yeah I guess it depends on what we want design wise right? If we want people to make entirely custom "games" as mods then it might make sense not to have any limitations on what a base set of animations are.

It also might be nice for prototyping fighter definitions/development workflow purposes to be able to load them with only a subset of animations but I think thats a different issue.

I guess we should note this as something to come back to. Definitely don't need to change that for current PR.